File: .codevalid/tests/task_5124983147_20260211210101/service_jsonplaceholder_fetch_users.py
import unittest
from unittest.mock import patch, MagicMock

import sys
import os

# Ensure the app directory is in the path for import
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../app/services')))
from app.services import jsonplaceholder_service

class TestJsonplaceholderServiceFetchUsers(unittest.TestCase):
    def setUp(self):
        self.service = jsonplaceholder_service.JsonplaceholderService

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._set_cached_users')
    def test_fetch_users_cache_expired_success_200(self, mock_set_cache, mock_get, mock_cache_expired):
        # Given
        users = [{"id": 1, "name": "User1"}, {"id": 2, "name": "User2"}]
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = users
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        mock_set_cache.assert_called_once_with(users)
        self.assertEqual(result, users)

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    def test_fetch_users_cache_expired_api_error_404(self, mock_get, mock_cache_expired):
        # Given
        mock_response = MagicMock()
        mock_response.status_code = 404
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    def test_fetch_users_cache_expired_api_error_500(self, mock_get, mock_cache_expired):
        # Given
        mock_response = MagicMock()
        mock_response.status_code = 500
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._set_cached_users')
    def test_fetch_users_cache_expired_api_empty_list(self, mock_set_cache, mock_get, mock_cache_expired):
        # Given
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = []
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        mock_set_cache.assert_called_once_with([])
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    def test_fetch_users_cache_expired_api_invalid_json(self, mock_get, mock_cache_expired):
        # Given
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.side_effect = ValueError("Invalid JSON")
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=False)
    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._get_cached_users')
    def test_fetch_users_cache_not_expired_return_cached_users(self, mock_get_cache, mock_cache_expired):
        # Given
        cached_users = [{"id": 1, "name": "CachedUser"}]
        mock_get_cache.return_value = cached_users

        # When
        result = self.service.fetch_users()

        # Then
        mock_get_cache.assert_called_once()
        self.assertEqual(result, cached_users)

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=False)
    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._get_cached_users')
    def test_fetch_users_cache_not_expired_cached_users_empty(self, mock_get_cache, mock_cache_expired):
        # Given
        mock_get_cache.return_value = []

        # When
        result = self.service.fetch_users()

        # Then
        mock_get_cache.assert_called_once()
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get', side_effect=Exception("Timeout"))
    def test_fetch_users_cache_expired_api_timeout(self, mock_get, mock_cache_expired):
        # Given: requests.get will raise an Exception (simulate timeout)

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get', side_effect=Exception("Connection error"))
    def test_fetch_users_cache_expired_api_connection_error(self, mock_get, mock_cache_expired):
        # Given: requests.get will raise an Exception (simulate connection error)

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        self.assertEqual(result, [])

    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._cache_expired', return_value=True)
    @patch('app.services.jsonplaceholder_service.requests.get')
    @patch('app.services.jsonplaceholder_service.JsonplaceholderService._set_cached_users')
    def test_fetch_users_cache_expired_api_partial_user_list(self, mock_set_cache, mock_get, mock_cache_expired):
        # Given
        partial_users = [{"id": 1}, {"name": "User2"}]  # missing fields
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = partial_users
        mock_get.return_value = mock_response

        # When
        result = self.service.fetch_users()

        # Then
        mock_get.assert_called_once_with("https://jsonplaceholder.typicode.com/users", timeout=5)
        mock_set_cache.assert_called_once_with(partial_users)
        self.assertEqual(result, partial_users)

if __name__ == '__main__':
    unittest.main()

File: .codevalid/tests/task_5124983147_20260211210101/service_jsonplaceholder_fetch_users.rb
require 'minitest/autorun'
require 'mocha/minitest'
require_relative '../../../app/services/jsonplaceholder_service'
require 'rails'

class ServiceJsonplaceholderFetchUsersTest < Minitest::Test
  def setup
    @cache = mock('cache')
    Rails.stubs(:cache).returns(@cache)
    @logger = mock('logger')
    Rails.stubs(:logger).returns(@logger)
    @logger.stubs(:error)
  end

  def mock_response(code, body)
    resp = mock('response')
    resp.stubs(:code).returns(code)
    resp.stubs(:body).returns(body)
    resp
  end

  # Test Case 1
  def test_fetch_users_cache_expired_success_200
    users = [{ 'id' => 1 }]
    @cache.expects(:fetch).with('jsonplaceholder_users', expires_in: JsonplaceholderService::CACHE_DURATION).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', users.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal users, result
  end

  # Test Case 2
  def test_fetch_users_cache_expired_api_error_404
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('404', ''))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 3
  def test_fetch_users_cache_expired_api_error_500
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('500', ''))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 4
  def test_fetch_users_cache_expired_api_empty_list
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', [].to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 5
  def test_fetch_users_cache_expired_api_invalid_json
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', 'invalid'))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 6
  def test_fetch_users_cache_not_expired_return_cached_users
    users = [{ 'id' => 1 }]
    @cache.expects(:fetch).returns(users)
    result = JsonplaceholderService.fetch_users
    assert_equal users, result
  end

  # Test Case 7
  def test_fetch_users_cache_not_expired_cached_users_empty
    @cache.expects(:fetch).returns([])
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 8
  def test_fetch_users_cache_expired_api_timeout
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).raises(Timeout::Error)
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 9
  def test_fetch_users_cache_expired_api_connection_error
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).raises(SocketError)
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 10
  def test_fetch_users_cache_expired_api_partial_user_list
    list = [{ 'id' => 1 }, { 'name' => 'Bob' }]
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', list.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal list, result
  end

  # Test Case 11
  def test_fetch_users_api_204_no_content
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('204', ''))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 12
  def test_fetch_users_cache_write_failure
    users = [{ 'id' => 1 }]
    @cache.expects(:fetch).raises(StandardError).then.yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', users.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal users, result
  end

  # Test Case 13
  def test_fetch_users_api_returns_non_array_json
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', { foo: 'bar' }.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 14
  def test_fetch_users_api_slow_but_not_timeout
    users = [{ 'id' => 1 }]
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', users.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal users, result
  end

  # Test Case 15
  def test_fetch_users_cache_contains_invalid_data
    @cache.expects(:fetch).returns('invalid_data')
    result = JsonplaceholderService.fetch_users
    assert_equal 'invalid_data', result
  end

  # Test Case 16
  def test_fetch_users_cache_key_missing
    users = [{ 'id' => 2 }]
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', users.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal users, result
  end

  # Test Case 17
  def test_fetch_users_api_returns_non_json_body
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', 'plain text'))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 18
  def test_fetch_users_api_redirect_301
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('301', ''))
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end

  # Test Case 19
  def test_fetch_users_api_large_payload
    large_list = Array.new(1000) { |i| { 'id' => i } }
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).returns(mock_response('200', large_list.to_json))
    result = JsonplaceholderService.fetch_users
    assert_equal large_list, result
  end

  # Test Case 20
  def test_fetch_users_api_ssl_error
    @cache.expects(:fetch).yields
    Net::HTTP.expects(:get_response).raises(OpenSSL::SSL::SSLError)
    result = JsonplaceholderService.fetch_users
    assert_equal [], result
  end
end
